lib/Backends/NNPI/NNPIMLTraceWrapper.cpp

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright (c) Glow Contributors. See CONTRIBUTORS file.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "NNPIMLTraceWrapper.h"
+#include "DebugMacros.h"
+#include <chrono>
+#include <cstring>
+#include <fstream>
+#include <memory>
+#include <sstream>
+#include <time.h>
+#include <unistd.h>
+#include <unordered_map>
+#include <vector>
+
+#define MAX_TRACE_BUFFER_SIZE (1024 * 1024 * 5)
+#define TRACE_READ_BUFFER_SIZE (1024 * 10)
+
+static inline uint64_t secondsToMicrosecons(double seconds) {
+  return (uint64_t)(seconds * 1e6f);
+}
+
+static inline int64_t nanosecondsToMicroseconsSigned(int64_t nanoseconds) {
+  return nanoseconds / 1e3;
+}
+
+static uint64_t getUpRefTime() {
+  static double refTime = 0;
+  if (refTime > 0) {
+    return refTime;
+  }
+  // Get uptime.
+  double uptime(0.0);
+  if (std::ifstream("/proc/uptime", std::ios::in) >> uptime) {
+    return secondsToMicrosecons(uptime);
+  }
+  return 0.0;
+}
+
+class NNPITraceParser {
+public:
+  NNPITraceParser(uint64_t timeDiffSec, uint64_t upRefTime)
+      : timeDiff_(timeDiffSec), upRefTime_(upRefTime), refTimeDiff_(0.0){};
+
+  int64_t getTimeDiffSec() {
+    return abs(refTimeDiff_) > 0 ? refTimeDiff_ : timeDiff_;
+  };
+
+  void parseLine(std::string line, NNPITraceEntry &entry) {
+    size_t idx = 0;
+    std::istringstream linestream(line);
+    do {
+      std::string part;
+      linestream >> part;
+
+      switch (idx) {
+      case 0: // Pid.
+      {
+        entry.processID = getPID(part);
+        break;
+      }
+      case 1: // CPU.
+      {
+        entry.cpuID = getCPUID(part);
+        break;
+      }
+      case 2: // Flags.
+      {
+        getFlags(part, entry.flags_);
+        break;
+      }
+      case 3: // TIMESTAMP.
+      {
+        entry.deviceUpTime = getOriginTime(part);
+        break;
+      }
+      case 4: // FUNCTION.
+      {
+        entry.traceType = getType(part);
+        break;
+      }
+      case 5: // May also be function for NNPI_TRACE_MARK.
+      {
+        if (entry.traceType == NNPI_TRACE_MARK &&
```

Comment:
Little confused by this -- isn't `entry` always going to be default constructed here to `NNPI_TRACE_UNKNOWN`, at least by current use cases (i.e. called from `NNPITraceContext::load()`)? Perhaps will be used differently in the future?

Reply:
@jfix71 Point taken. Added a comment to make it a little clearer and will revisit the parsing in next versions.